### PR TITLE
Shonali lux layer

### PR DESCRIPTION
**Pull request by shonalidixit**
Description

This PR introduces an initial Lux-based implementation for directed hypergraph neural networks.

Summary of changes
- Added a new `layers/` directory to organise neural network layers.
- Implemented a reusable `DirectedHypergraphLayer` using Lux.jl.
- Added an initial `DirectedHypergraphRegression` model built on top of the directed hypergraph layer.
- Integrated the new layers into the package.
- Added basic tests for layer initialisation and the forward pass.

This PR is intended as an initial implementation for review and discussion before further development.

### CODE REVIEW

File: src/layers/DirectedHypergraphLayer.jl [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+zero while leaving those columns unchanged.
+"""
+
+function safe_column_normalise(M::AbstractMatrix)
+    col_sums = sum(M, dims = 1)
+    safe_sums = similar(col_sums)
+
+    for i in eachindex(col_sums)
+        safe_sums[i] = col_sums[i] == 0 ? one(eltype(col_sums)) : col_sums[i]
+    end
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphLayer.jl]: Why not combine this with `safe_row_normalize` and just have the dimension be an input parameter?

File: src/layers/DirectedHypergraphLayer.jl [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+function safe_row_normalise(M::AbstractMatrix)
+    row_sums = sum(M, dims = 2)
+    safe_sums = similar(row_sums)
+
+    for i in eachindex(row_sums)
+        safe_sums[i] = row_sums[i] == 0 ? one(eltype(row_sums)) : row_sums[i]
+    end
+
+    return M ./ safe_sums
+end
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphLayer.jl]: See above.

File: src/layers/DirectedHypergraphLayer.jl [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+"""
+    DirectedHypergraphLayer(species_in_dim, hidden_dim, activation)
+
+A Lux-compatible message-passing layer for directed hypergraphs.
+
+The layer accepts a species-feature matrix together with source and target
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphLayer.jl]: "species" is specific to CRNs. Remember that `HyperGraphNeuralNetworks.jl` is a generic hypergraph ML library.

File: src/layers/DirectedHypergraphLayer.jl [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+incidence matrices. It performs:
+
+1. A learnable transformation of species features.
+2. Separate aggregation of source and target species into reaction embeddings.
+3. A learnable transformation of reaction embeddings.
+4. Propagation of reaction messages back to participating species.
+5. A learnable update of the species embeddings.
+
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphLayer.jl]: See above. The layers that you develop should be general to any (di)hypergraph.

File: src/layers/DirectedHypergraphLayer.jl [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+# Arguments
+
+- `species_in_dim`: Number of input features associated with each species.
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphLayer.jl]: Should be something like `vertex_in_dim`. You should also have a `hyperedge_in_dim`, I would think.

File: src/layers/DirectedHypergraphLayer.jl [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+- `reaction_embeddings`: Learned reaction embeddings.
+"""
+
+struct DirectedHypergraphLayer{F} <: Lux.AbstractLuxLayer
+    species_in_dim::Int
+    hidden_dim::Int
+    activation::F
+end
+
+function Lux.initialparameters(
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphLayer.jl:157]: Would be nice if you allowed the user to input various methods of weight initialization. See https://github.com/JuliaGraphs/GraphNeuralNetworks.jl/blob/master/GNNLux/src/layers/conv.jl

File: src/layers/DirectedHypergraphLayer.jl [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+            2 * layer.hidden_dim,
+            layer.hidden_dim
+        ) .* 0.1f0,
+
+        b_reaction = zeros(
+            Float32,
+            1,
+            layer.hidden_dim
+        ),
+
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphLayer.jl]: Why are biases always initialized to zero?

File: src/layers/DirectedHypergraphLayer.jl [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            1,
+            layer.hidden_dim
+        )
+    )
+end
+
+Lux.initialstates(
+    ::AbstractRNG,
+    ::DirectedHypergraphLayer
+) = NamedTuple()
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphLayer.jl:235]: Should this be empty? Shouldn't the NN state be part of this?

File: src/layers/DirectedHypergraphLayer.jl [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+"""
+
+function (layer::DirectedHypergraphLayer)(input, ps, st)
+    X_species, source_matrix, target_matrix = input
+
+    membership_matrix = source_matrix .+ target_matrix
+
+    source_norm = safe_column_normalise(source_matrix)
+    target_norm = safe_column_normalise(target_matrix)
+    membership_norm = safe_row_normalise(membership_matrix)
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphLayer.jl]: Will the user always want to normalize?

File: src/layers/DirectedHypergraphRegression.jl [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+#combines a DirectedHypergraphLayer with a Lux Dense regression head.
+#the model produces one continuous prediction for every reaction/hyperedge.
+
+
+struct DirectedHypergraphRegression{H, R} <:
+       Lux.AbstractLuxContainerLayer{(:hypergraph_layer, :regression_head)}
+
+    hypergraph_layer::H
+    regression_head::R
+end
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphRegression.jl]: Again, think about your end user. Is this something that an average end user will want, or will they be designing their own network architectures? I think likely the latter. `HyperGraphNeuralNetworks.jl` should (mainly) be general-purpose features that could be used for HGNNs. We shouldn't be pigeon-holing folks into one particular architecture.

File: src/layers/DirectedHypergraphRegression.jl [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+→ dense regression head
+→ one scalar prediction per reaction
+
+# Arguments
+
+- `species_in_dim`: Number of input features for each species.
+- `hidden_dim`: Size of the hidden embeddings.
+- `activation`: Activation function used in the hypergraph layer.
+"""
+function DirectedHypergraphRegression(
```
Comment (espottesmith) [inline src/layers/DirectedHypergraphRegression.jl]: Here, you're assuming a single message-passing layer, we we discussed. At very least, you should have a variable number of layers.

File: test/layers/DirectedHypergraphLayer.jl [hunk 12]
```diff
@@ -0,0 +1,47 @@
+using Test
```
Comment (espottesmith) [inline test/layers/DirectedHypergraphLayer.jl:1]: Personal style, but unless your tests are massive, I prefer to have one test file with multiple `@testset`s

File: test/layers/DirectedHypergraphLayer.jl [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        1 0 0;
+        0 0 1
+    ]
+
+    output, st = layer(
+        (X_species, source_matrix, target_matrix),
+        ps,
+        st
+    )
+
```
Comment (espottesmith) [inline test/layers/DirectedHypergraphLayer.jl:408]: You're testing that the dimensions are what you expect and that you don't get any exploding infinite numbers. Is that all you need?

File: test/runtests.jl [hunk 14]
```diff
@@ -9,6 +9,7 @@ using MLUtils
 using SimpleHypergraphs
 using SimpleDirectedHypergraphs
 using HyperGraphNeuralNetworks
+include("layers/DirectedHypergraphLayer.jl")
```
Comment (espottesmith) [inline test/runtests.jl:12]: If you're exporting properly, you shouldn't need to do this.

**Comment by espottesmith** [comment]
Solid improvement on the test. Unless I missed it, you're still not testing gradients/differentiation (which is important for backpropagation and ML). Take a look at how `GraphNeuralNetworks.jl` runs their tests, e.g., https://github.com/JuliaGraphs/GraphNeuralNetworks.jl/blob/f80bc29ba47a987cbc69b4563e5109b0c17ad18d/GNNLux/test/test_module.jl

The `DirectedHypergraphLayer` should be renamed (there will presumably be a number of directed hypergraph layers). Will take another close look later.

Also, for the final PR, please remove your sandbox files (i.e., all of your numbered files, like examples/shonali_prototypes/12_formose_crn_case_study.jl). These shouldn't be part of the main codebase.
